### ISSUE TITLE
Proposal: Always recommend dedicated Kibana instance for dedicated Monitoring ES cluster

### ISSUE THREAD

**Issue by ycombinator** [elastic/stack-monitoring#37]
Currently, when it comes to setting up Stack Monitoring, three setups are possible:

1. Using the Production Elasticsearch cluster as the Monitoring Elasticsearch cluster as well (i.e. using the `local` exporter) and looking at the Stack Monitoring UI in the Production Kibana instance (i.e. setting `elasticsearch.*` but not `xpack.monitoring.elasticsearch.*` in `kibana.yml`).

2. Using a dedicated, remote Monitoring Elasticsearch cluster that's separate from the the Production Elasticsearch cluster (i.e. using the `http` exporter) and looking at the Stack Monitoring UI in the Production Kibana instance (i.e. setting `xpack.monitoring.elasticsearch.*` in `kibana.yml`).

3. Using a dedicated, remote Monitoring Elasticsearch cluster that's separate from the the Production Elasticsearch cluster (i.e. using the `http` exporter) and looking at the Stack Monitoring UI in the dedicated Monitoring Kibana instance that's separate from the Production Kibana instance (i.e. setting `elasticsearch.*` but not `xpack.monitoring.elasticsearch.*` in the Monitoring Kibana instance's `kibana.yml`).

## Current State

Today, for the **Getting Started** use case, we recommend setup 1.

For the **Production** use case, we recommend setup 2. And we also say that users ["might want to"](https://www.elastic.co/guide/en/x-pack/current/how-monitoring-works.html) use setup 3.

## Proposal

For the **Production** use case, we should only recommend setup 3 and stop recommending setup 2.

### Pros
* Truly isolates the monitoring stack from the production stack(s). This is more resilient to failures in the production stack(s). Users will be able to monitoring their production stack(s) even if one of them is severely unresponsive.

* Monitoring-related Kibana apps like the Logs Viewer do not need to know how to talk to a Monitoring Elasticsearch cluster (i.e. the one specified via `xpack.monitoring.elasticsearch.*` in `kibana.yml`). They can continue to talk to Kibana's Production Elasticsearch cluster (i.e. the one specified via `elasticsearch.*` in `kibana.yml`). Related: https://github.com/elastic/kibana/issues/31657.

* TBD (@cachedout, @chrisronline, @yaronp68: I forgot what this was. If you remember, could you edit and populate this bullet point, please?)

* Makes implementing non-native realm Security easier. (@kobelb could you summarize here why this would be the case?)

### Cons

* Requires additional setup, specifically the setup of an additional Kibana instance dedicated for Monitoring.

**Comment by chrisronline** [comment]
I want to address some offline feedback to one of the cons listed, where we lose being able to monitor and manage from a single Kibana instance.

I think we need to be more honest with ourselves about the ability to manage and monitor a cluster, in terms of where we are and where we can possibly go.

Right now, Kibana is only able to manage (perform write actions) a single cluster - the cluster it knows about it. It's not able to perform any write actions against any other cluster. It can also read from another cluster - a dedicated monitoring cluster.

If we continue recommending option 2, nothing changes about the previous statement. A single Kibana instance can read and write to a single production ES cluster, and read from a single dedicated monitoring cluster. This is an ideal setup in that users are able to monitor and manage _from a single location_ (the single Kibana instance). This is really what we want - these two things should co-exist as there is so much more value gain from not only being able to identify the problem, but also fix it.

But this immediately falls short when you consider any multi-cluster environment. As soon as there is a second ES cluster, a single Kibana instance isn't able to read and write to _both_ ES clusters anymore. It has to choose one and if the user wants to be able to read and write to the other, they need to spin up a new Kibana instance. When we consider this at scale, does it become a nonstarter? 

Option 2 works fine in a single cluster environment, but it doesn't seem to scale for multi-cluster setups. How do we intend to solve this problem for multi cluster environments? I don't know if we've had ideas thrown around for this, but we should link them here if we can find them.

**Comment by cachedout** [comment]
There was a brief discussion in Slack about this but I wanted to add some follow-up points.

@skearns64 made some excellent points about the UX implications of this decision. I wouldn't want to restate his position here so I'll leave that to him. :)

I did want to heavily emphasize, though, that while I agree that from a user perspective that it is convenient to be able to view monitoring data from the production Kibana instance, that doing so comes at an enormous cost, which is that, in doing so, _monitoring remains almost wholly dependent on the services we are monitoring_. [Point 1 in the "Pros" list.] 

Historically (as I understand it, at least) decoupling monitoring from the production services is one of the primary drivers -- if not the greatest driver -- behind the major effort to move to a dedicated monitoring cluster to begin with. 

IMHO, given that we're already settled on setting up additional instances of Elasticsearch for monitoring as the recommended path forward, it does not seem particularly burdensome to set up a Kibana instance alonside it. [Point 2 in the "Cons list"] and the advantages to the user are clear, though we would certainly do well to explain them in a detailed way in the setup documentation.

In my experience thus far, Kibana is *heavily* dependent on the health of the cluster that it is attached to. While the stack monitoring application itself may have what it needs to contact the monitoring cluster, what good does this do if the Kibana instance upon which it depends becomes seriously degraded?

From a useability perspective, in putting the Stack Monitoring application on the Production Kibana instance, we very much risk putting our users in a situation where, at the point where a _user needs monitoring most_ -- to quickly diagnose and understand a degraded cluster -- it will be unavailable.

If we decide not to recommend a separate Kibana instance, I believe we'll need to do additional testing and work to try to figure out how Stack Monitoring truly behaves when  Kibana's Elasticsearch production cluster becomes unresponsive and how we might make Kibana resilient enough to keep the Stack Monitoring application functional in those cases.

**Comment by cachedout** [comment]
I would like to gather some consensus on this issue, as it is currently blocking a planned major feature.

@skearns64 Do you have additional insight on this you would like to share? Or, @monicasarbu or @yaronp68 ? I am happy to schedule a meeting in order to discuss this comprehensively but wanted to give folks a chance to read through the thread and think things over before that. :) Thanks!

**Comment by Dumbaz** [comment]
In multi cluster setups, there is an additional issue with setup 2.

During the time the upgrade is performed, the different versions of ES will prevent one from using Kibana for the clusters which do not have the same version as your Kibana instance.

Scenario: You have two separate clusters, a production cluster  and a dedicated monitoring cluster. One Kibana Instance is used two access data from both clusters. During the Upgrade to 6.6 of the production cluster Kibana is still usable for the monitoring data as long as it remains on 6.5. Once it is upgraded to 6.6 it does no longer work for monitoring but will work for production data again. Full functionality is restored once everything is on the same version.

Completely separate stacks (ES + Kibana) prevent this issue.

**Comment by yaronp68** [comment]
CC: @bmcconaghy 
I propose to treat the monitoring cluster + kibana as a monitoring service. The monitoring service supply the following functionality:

1. Index and store monitoring data (including retention policy etc)
2. The monitoring service performs search, aggregations and data formatting for the management plugins that needs monitoring data (e.g /monitor/nodes and in the body you mention some built in filters like data nodes by index speed)
3. The monitoring GUI uses the monitoring service API to provide insights, visualisations and navigation over the monitoring data

The monitoring GUI can be ran on the monitoring kibana and show all clusters or run over the prod Kibana and show only single cluster data

I see the following benefits:
- No direct access to monitoring ES by any other Kibana
- The API provides processed data and can potentially provide some caching to take the load off ES
- The API provides security based on Kibana security and uses under the hood Elasticsearch data level security or filters to make sure only permitted data is shown based on client

**Comment by bmcconaghy** [comment]
We would like to integrate monitoring data into management screens (index management in particular, but also maybe others) so whatever solution we reach here I would hope that it doesn't preclude this.  I would also hope that we can do the reverse: expose some management features in the monitoring screens, such that people could fix issues that monitoring might point out to them.

**Comment by cachedout** [comment]
There have been good points raised. I'll address them below:

**Displaying monitoring data in management interfaces**
I agree that this makes a lot of sense. One thing that we've talked about in the past is making it easier for components from stack monitoring to be embedded in other interfaces, such as Kibana's management screen(s). I think that we can likely achieve this regardless of whether or not the stack monitoring application itself is local or remote. @chrisronline do you agree?

**Allowing management actions from monitoring**
I presume that the sort of actions that are being imagined are those which could be instantiated through calls to the Elasticsearch REST API. If so, I am favor of this but believe that we should proceed with caution. If we go down this road with a dedicated Kibana instance for monitoring, we cross the threshold into a world where stack monitoring is now a single point from which an individual could potentially control multiple clusters. There are certainly ways in which this is desirable but we'll have to think very carefully about security and permissions. For this reason, I have some hesitation about this path.

Here, I think, we also have to consider the comments from @chrisronline above. @chrisronline with your statement, `a single Kibana instance isn't able to read and write to both ES clusters anymore`, does this mean that if we had a dedicated monitoring Kibana that it would not be able to manage (i.e. write to) multiple monitored clusters in order to manage them?

I still strongly believe that we should go with option 3 in the list that @ycombinator has created but also think that we should have a clear statement about how this might affect the remote management capabilities that both @yaronp68 and @bmcconaghy are advocating for here.

**Comment by chrisronline** [comment]
>  I think that we can likely achieve this regardless of whether or not the stack monitoring application itself is local or remote. @chrisronline do you agree?

I think so yes. We'd just need to make sure that the management (or whoever is doing the embedding) can handle any potential security/authorization issues around connecting to the dedicated monitoring cluster. We currently have a separate connection established for the monitoring cluster but I'm not sure how that will change if we start recommending users to use a dedicated Kibana for the dedicated monitoring cluster (since we wouldn't need the separate connection at that point)

> with your statement, a single Kibana instance isn't able to read and write to both ES clusters anymore, does this mean that if we had a dedicated monitoring Kibana that it would not be able to manage (i.e. write to) multiple monitored clusters in order to manage them?

We currently can't manage more than one cluster. We are limited to managing the cluster that Kibana knows about (whatever is configured through `elasticsearch.hosts`). In our current world, the Kibana stack monitoring app _could_ make writes to this cluster, but if we recommend a dedicated Kibana for the dedicated monitoring cluster, that configuration (`elasticsearch.hosts`) will point to the dedicated monitoring cluster so we could only make writes to that cluster (which I'm assuming will require less management than production clusters). Either way, we can only manage a single instance and this has been true for some time.

So, with this limitation in mind, how do we move from where we are to being able to "manage multiple clusters" from a single Kibana? Is that a goal we want, and if so, do we have any known, tracked work items to get us there? 

If we don't want to move to that approach (and we're happy with the single cluster manage limitation), what are we _really_ losing when switching to a dedicated Kibana for the dedicated monitoring cluster? I know there is some functionality we'd lose from the start (for example, we do some logic to determine the "main" cluster in a list of clusters to enable certain functionality - we identify the "main" cluster as the one Kibana is configured to (`elasticsearch.hosts`)), but are we losing enough to justify making the switch? Can we keep some of this functionality, even after switching?

**Comment by cachedout** [comment]
> If we don't want to move to that approach (and we're happy with the single cluster manage limitation), what are we really losing when switching to a dedicated Kibana for the dedicated monitoring cluster?

Thanks, @chrisronline for the excellent overview!

As I understand this, we wouldn't be losing anything that we have _currently_ but recommending a dedicated Kibana instance could potentially limit our ability to introduce management capability into the stack monitoring application running on that instance in the _future_.

My view here is that losing the ability to manage (multiple) production systems from a monitoring interface is an acceptable tradeoff for drastically increasing the isolation of a monitoring system. While I understand that giving an individual the ability to manage a system from a monitoring application is desirable from a feature perspective, I'd point out that we can potentially include some monitoring data in the management pages of production Kibanas. So, we could accomplish one half of the scenario that @bmcconaghy has laid out.

To be honest, if we want to build that I would be more in favor of opening a discussion about building a centralized _management_ application that could live alongside Stack Monitoring  on this (or another) dedicated cluster than I would trying to co-mingle management capabilities inside monitoring.

I believe that would be a much cleaner solution versus trying to inject management capability into the stack monitoring application. It cleanly sidesteps the concerns about compartmentalizing management and monitoring from a user authorization perspective and, in the near term, I don't see the need to pull up  a production Kibana instance in order to make changes to that instance as one that will be at all burdensome or surprising to users.

Comments/criticisms/rebuttals welcome as always. :)

**Comment by yaronp68** [comment]
@cachedout Let's focus on the scenario of embedding monitoring information (e.g. lagging behind CCR followers with the stats that show that) into management and ignore completely the other direction monitoring -> management action which is much more controversial. There are several options here. I suggest we choose one of them

**Comment by cachedout** [comment]
@yaronp68 Thanks for the swift response. Given the discussion thus far, I would suggest that we leave this issue open for another twenty-four hours in case there are comments yet to come in. Barring any objection in that time period, it seems clear that option 3 is the direction that we'd like to go in.

**Comment by skearns64** [comment]
@cachedout - apologies for missing the 24 hour deadline - would you mind clarifying what "option 3" is and what going in that direction entails? I'm not confident that I follow what we will be changing as a result of a decision here.

> it is currently blocking a planned major feature
 
Which feature is this blocking?

**Comment by chrisronline** [comment]
> Which feature is this blocking?

It's not currently blocking any feature. We had originally thought it was holding up the Logs UI PR, but we've decided to solve the problem a different way. [Read more about that here](https://github.com/elastic/kibana/issues/31657#issuecomment-472980906)
